Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,283 @@
+            throw BSONDecimal128Errors.Rounding
+        }
+
+        guard digits.count <= BSONDecimal128.PERCISION_DIGITS else {
+            throw BSONDecimal128Errors.Overflow
+        }
+
+        var significand = UInt128(hi: 0x0, lo: 0x0)
+
+        if digits.isEmpty {
+            significand.hi = 0
+            significand.lo = 0
+        }
+
+        if digits.count < 17 {
+            significand.lo = UInt64(digits[0])
+            for digit in digits[1...] {
+                significand.lo *= 10
+                significand.lo += UInt64(digit)
+            }
+        } else if digits.count > 17 {
+            significand.hi = UInt64(digits[0])
+            for digit in digits[1..<18] {
+                significand.hi *= 10
+                significand.hi += UInt64(digit)
+            }
+
+            significand.lo = UInt64(digits[18])
+            for digit in digits[19...] {
+                significand.lo *= 10
+                significand.lo += UInt64(digit)
+            }
+        }
+
+        self.value = UInt128(hi: 0, lo: significand.lo)
+        let biased_exponent = exponent + BSONDecimal128.EXPONENT_BIAS
+
+        // Encode combination, exponent, and significand.
+        if (significand.hi >> 49) & 1 == 1 {
+            // Encode '11' into bits 1 to 3
+            self.value.hi |= (0b11 << 61)
+            self.value.hi |= UInt64(biased_exponent & 0x3FFF) << 47
+            self.value.hi |= significand.hi & 0x7FFF_FFFF_FFFF
+        } else {
+            self.value.hi |= UInt64(biased_exponent & 0x3FFF) << 49
+            self.value.hi |= significand.hi & 0x1_FFFF_FFFF_FFFF
+        }
+
+        if sign < 0 {
+            self.value.hi |= 0x8000_0000_0000_0000
+        }
+    }
+
+    private static func convertToDigitsArray(_ decimalString: String) throws -> [UInt8] {
+        var leadingZero = true
+        var digits: [UInt8] = []
+
+        let asciiZero = Character("0").asciiValue ?? 48
+        let asciiNine = Character("9").asciiValue ?? 57
+        let asciiPoint = Character(".").asciiValue ?? 46
+
+        let digitsFromRepr = [UInt8](decimalString.utf8)
+        if digitsFromRepr.count > 1 {
+            for digit in digitsFromRepr {
+                if digit == asciiPoint {
+                    continue
+                }
+                if (digit < asciiZero) && (digit > asciiNine) {
+                    throw BSONDecimal128Errors.Syntax
+                }
+                if digit == asciiZero && leadingZero {
+                    continue
+                }
+                if digit != asciiZero && leadingZero {
+                    // seen a non zero digit
+                    leadingZero = false
+                }
+                digits.append(digit - asciiZero)
+            }
+        } else {
+            digits.append(digitsFromRepr[0] - asciiZero)
+        }
+        return digits
+    }
+
+    public static func read(from buffer: inout ByteBuffer) throws -> BSON {
```

Comment:
since the main type definition is already pretty long, consider moving the BSONValue conformance into a extension in the same file just to make things a bit more organized, like

```swift
extension BSONDecimal128: BSONValue {
   // methods here
}
```